### DISCUSSION TITLE
Architecture of lithium

### DISCUSSION THREAD

**Discussion by shivshankardayal** [matt-42/lithium#61] [General]
It would be nice to have the architecture of lithium presented in the documentation page. Right now it is not clear as to how it works internally.

**Comment by matt-42** [comment]
Hi @shivshankardayal ,

Thanks for your feedback. We can do that yes. Can you tell more precisely what you would like to see in this "internal architecture" section ? and why do you need it ? I'm asking because "internal architecture" can mean lots of things.

**Reply by shivshankardayal** [reply]
What I would like to see is after a request is arrived how `lithium` handles ii? In case of high concurrency how coroutines behave? How stack memory is managed in case of high concurrency?

**Reply by shivshankardayal** [reply]
Also, is there a plugin architecture in place for `lithium`?

**Reply by matt-42** [reply]
Hi @shivshankardayal , 

Here is more info about the internals of lithium : 
Everytime a new http connection gets in, lithium starts a continuation (using boost context callcc [1]) that processes the http requests sent by this connection. callcc allocates one stack by continuation and manages it internally, (more info in [1]), so three is no code that explicitely deal with the stack in lithium.


[1] https://www.boost.org/doc/libs/1_74_0/libs/context/doc/html/context/cc.html

**Reply by matt-42** [reply]
> Also, is there a plugin architecture in place for `lithium`?

What do you mean by plugin ? (I'm asking because I know people use this word for lots of different things).

**Reply by shivshankardayal** [reply]
Thanks. It would be much better to put this information in docs site.

**Reply by shivshankardayal** [reply]
Plugin like Wordpress plugins like there can be a plugin to send email or a client to a DB server.

**Reply by matt-42** [reply]
Hi @shivshankardayal ,

We don't need plugins to send emails with lithium. Just a new library that send emails.
If we add emailing is added one day into lithium, it will be as a new library with functions to send mails.
